### ISSUE TITLE
`podman image scp` does not use installed SSH binary meaning SSH configuration does not work.

### ISSUE THREAD

**Issue by igo95862** [containers/podman#11963]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

**Steps to reproduce the issue:**

1. Add a new host to `$HOME/.ssh/config`

2. Try `podman image scp` to copy image to new host

**Describe the results you received:**

Host not found.

`Error: failed to connect: dial tcp: lookup newhost: no such host`

**Describe the results you expected:**

Successfully works just like `rsync`, `git`, `sshfs`...

**Additional information you deem important (e.g. issue happens only occasionally):**

It uses `golang.org/x/crypto/ssh` native golang SSH implementation. In contrast `rsync`, `git` and `sshfs` uses the installed binary and communicates over SSH stdin and stdout.

It should be possible to implement an SSH client over external binary which would give a benefit of installed SSH client configuration.

https://github.com/containers/image/issues/958

**Output of `podman version`:**

```
Version:      3.4.0
API Version:  3.4.0
Go Version:   go1.17.1
Git Commit:   6e8de00bb224f9931d7402648f0177e7357ed079
Built:        Fri Oct  1 13:14:18 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers: []
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: /usr/bin/conmon is owned by conmon 1:2.0.30-1
    path: /usr/bin/conmon
    version: 'conmon version 2.0.30, commit: 2792c16f4436f1887a7070d9ad99d9c29742f38a'
  cpus: 12
  distribution:
    distribution: arch
    version: unknown
  eventLogger: journald
  hostname: lumine
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65537
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65537
  kernel: 5.14.10-arch1-1
  linkmode: dynamic
  logDriver: journald
  memFree: 25989746688
  memTotal: 33571397632
  ociRuntime:
    name: crun
    package: /usr/bin/crun is owned by crun 1.2-1
    path: /usr/bin/crun
    version: |-
      crun version 1.2
      commit: 4f6c8e0583c679bfee6a899c05ac6b916022561b
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: /usr/bin/slirp4netns is owned by slirp4netns 1.1.12-1
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.2
  swapFree: 0
  swapTotal: 0
  uptime: 12m 8.8s
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries: {}
store:
  configFile: /home/igo95862/.config/containers/storage.conf
  containerStore:
    number: 28
    paused: 0
    running: 0
    stopped: 28
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: /usr/bin/fuse-overlayfs is owned by fuse-overlayfs 1.7.1-1
      Version: |-
        fusermount3 version: 3.10.5
        fuse-overlayfs: version 1.7.1
        FUSE library version 3.10.5
        using FUSE kernel interface version 7.31
  graphRoot: /home/igo95862/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 29
  runRoot: /run/user/1000/containers
  volumePath: /home/igo95862/.local/share/containers/storage/volumes
version:
  APIVersion: 3.4.0
  Built: 1633083258
  BuiltTime: Fri Oct  1 13:14:18 2021
  GitCommit: 6e8de00bb224f9931d7402648f0177e7357ed079
  GoVersion: go1.17.1
  OsArch: linux/amd64
  Version: 3.4.0

```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
Name            : podman
Version         : 3.4.0-1
Description     : Tool and library for running OCI-based containers in pods
Architecture    : x86_64
URL             : https://github.com/containers/podman
Licenses        : Apache
Groups          : None
Provides        : None
Depends On      : cni-plugins  conmon  containers-common  crun  fuse-overlayfs
                  iptables  libdevmapper.so=1.02-64  libgpgme.so=11-64
                  libseccomp.so=2-64  slirp4netns
Optional Deps   : apparmor: for AppArmor support
                  btrfs-progs: support btrfs backend devices [installed]
                  catatonit: --init flag support
                  podman-docker: for Docker-compatible CLI [installed]
Required By     : podman-docker
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 72,66 MiB
Packager        : David Runge <dvzrv@archlinux.org>
Build Date      : 2021-10-01T13:14:18 MSK
Install Date    : 2021-10-11T11:52:03 MSK
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature
Backup Files    :
(none)

```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

No

**Comment by rhatdan** [comment]
Sure, as long as we hide this from the user.  
@jwhonce WDYT?

**Comment by cdoern** [comment]
@jwhonce @rhatdan not sure if I am interpreting this correctly, if we allow for users to manually add and specify hosts, doesn't that kind of open the door for more issues in terms of verifying if a podman socket is active on the other end? Currently `image scp` takes either a connection added and verified via `podman system connection add` as well as new, full urls containing the path to the socket on the remote host. Even the direct connection is kind of iffy and could be messed with if used improperly. The tighter of a lid we keep on this, probably the better? Again, I might be overcomplicating this issue.

**Comment by igo95862** [comment]
I've experimented with combining Golang and OpenSSH.

I think the easiest way would be to share a unix domain socket using `-L` option. It is available since OpenSSH 6.7. (from 2014 so it should be available in all distros)

This is what waypipe does.

`net/rpc` golang package can then used to communicate between processes. (I think you can even use gRPC over unix domain sockets.)

As for socket name I originally thought about using UUID and mount it over /tmp however, one of the escape sequence to forward a new socket while connection is open so mounting some socket in XDG_RUNTIME_DIR should be possible.

**Comment by igo95862** [comment]
It turns out easier to run RPC over SSH. All I have to do is to use STDIN and STDOUT that are piped over SSH. The client only has to pass the commands that start the matching program to the SSH program.

I have put the library in to this repository.

https://github.com/igo95862/golang-rpc-over-ssh-cmd

I can start experimenting with integrating it with podman next.

**Comment by rhatdan** [comment]
@igo95862 @cdoern What is up with this issue?

**Comment by igo95862** [comment]
My SSH RPC library works but I did not have any feedback on it...

**Comment by cdoern** [comment]
@igo95862 I can make some changes for port assumptions if you want. I currently have a PR improving the syntax and error messages for SCP. Other than that I feel like this falls back to a decision from @baude or @jwhonce on supporting openSSH

**Comment by LewisGaul** [comment]
Similarly `podman-remote` seems to not respect user SSH config - this would certainly be nice to have.

**Comment by rhatdan** [comment]
The SSH config would be on the server side though, correct?

**Comment by igo95862** [comment]
> The SSH config would be on the server side though, correct?

Doesn't the client initiate the connection?

**Comment by rhatdan** [comment]
Yes, I misread.  We are using builtin ssh library from golang.

**Comment by cdoern** [comment]
I can *probably* wire this up to look at the ssh config file and make a new podman system connection from that... @containers/podman-maintainers WDYT? This could open some new use cases for scp and for podman remote in general

**Comment by rhatdan** [comment]
I don't believe that is the issue.  The problem as I understand it, is that the opessh library we are using for ssh does not fully support the contents of the users ssh config file.  We need to investigate if there is a better library to use, or a way to allow users to exec the ssh command on their platform rather then using the ssh library.

**Comment by igo95862** [comment]
Using SSH binary to connect will open up all features that OpenSSH can use.

For example, OpenSSH supports FIDO2/U2F hardware keys.

**Comment by rhatdan** [comment]
I think making it optional makes some sense.  The problem we are trying to avoid is the assumption that all ssh clients are the same, specifically on MAC and Windows. If the ssh executable is optional then it is the users problem if it does not work correctly.

**Comment by rhatdan** [comment]
@cdoern don't you have a PR for this?

**Comment by cdoern** [comment]
Yep. This is closed via #15094
